beacon_node/eth2_libp2p/src/rpc/mod.rs

```diff
@@ -193,13 +216,47 @@ where
         conn_id: ConnectionId,
         event: <Self::ProtocolsHandler as ProtocolsHandler>::OutEvent,
     ) {
-        // send the event to the user
-        self.events
-            .push(NetworkBehaviourAction::GenerateEvent(RPCMessage {
-                peer_id,
-                conn_id,
-                event,
-            }));
+        if let Ok(RPCReceived::Request(ref id, ref req)) = event {
+            // check if the request is conformant to the quota
+            match self.limiter.allows(&peer_id, req) {
+                Ok(()) => {
+                    // send the event to the user
+                    self.events
+                        .push(NetworkBehaviourAction::GenerateEvent(RPCMessage {
+                            peer_id,
+                            conn_id,
+                            event,
+                        }))
+                }
+                Err(RateLimitedErr::TooLarge) => {
+                    // we set the batch sizes, so this is a coding/config err
+                    crit!(self.log, "Batch too large to ever be processed";
+                        "protocol" => format!("{}", req.protocol()));
+                }
+                Err(RateLimitedErr::TooSoon(until)) => {
+                    warn!(self.log, "Request exceeds the rate limit";
+                        "request" => req.to_string(), "peer_id" => peer_id.to_string(), "earliest_in_ms" => until.as_millis());
+                    let event = Err(HandlerErr::Inbound {
+                        id: *id,
+                        proto: req.protocol(),
+                        error: RPCError::RateLimited,
+                    });
+                    self.events
```

Comment:
Can we also respond to the peer with an Error, telling them they got rate limited, also in the case `TooLarge`

Other clients were discussing sending back the `until` also, but I don't think we support that yet in the spec.

Reply:
I sent (as a string) how much they should wait. If we then need the actual until, it won't be hard